### ISSUE TITLE
backintime-qt script uses non-absolute python3 call

### ISSUE THREAD

**Issue by eddgrant** [bit-team/backintime#1316]
Hey folks,

This was baffling me for ages, but I had a flash of inspiration which led me to understand what was going on so I thought I'd share what I have found.

At some point recently the backintime GUI stopped working for me. I initially thought this might have been due to my recent upgrade to Ubuntu 22.04 LTS but none of the solutions I found helped my particular issue.

When I run  `/usr/bin/backintime-qt` I got the following error:

```shell
Traceback (most recent call last):
  File "/usr/share/backintime/qt/app.py", line 35, in <module>
    import qttools
  File "/usr/share/backintime/qt/qttools.py", line 21, in <module>
    from PyQt5.QtGui import (QFont, QColor, QKeySequence)
ModuleNotFoundError: No module named 'PyQt5'
```

My initial thought was that I was somehow missing the `PyQt5` package so I checked using `pip3 freeze | grep PyQt5` and it confirmed that I had `PyQt5` installed:

```shell
pip3 freeze | grep PyQt5
PyQt5==5.15.6
PyQt5-sip==12.9.1
```

I then looked at the contents of `/usr/bin/backintime-qt` and noticed that it has the `#!/bin/sh` shebang line. Furthermore when it calls out to python (`python3 -Es ${APP_PATH}/app.py "$@"`) it doesn't specify the path to the python binary.

I think what was happening in my case was that my [pyenv](https://github.com/pyenv/pyenv) configuration was being initialised in the process in which `backintime-qt` was being run. When it initialised pyenv inserts various "shims" in to the `PATH` environment variable, with the intention of providing access to a specific Python runtime. The `backintime-qt` script seems to be expecting to use the _system_ python runtime.

To test this I replaced the line which calls out to python with the following, which ensures that it calls my system python runtime:

```shell
/usr/bin/python3 -Es ${APP_PATH}/app.py "$@"
```

Lo and behold this resolves the issue and the backintime GUI now works for me again.

I'm not sure what a "proper" fix might look like here, as I'm aware that hard coding the path to the python runtime might not be portable across all Linux distros. I did wonder if there was a way to ensure the `backintime-qt` script was run in a "non-interactive" way, in the hope this might avoid `pyenv` and other similar tools from polluting backintime's environment, but I didn't have any luck when testing this.

Anyway, thought this was worth sharing and hope this helps someone.

P.S. BackInTime is a fabulous tool, much better than Ubuntu's default backup tool. Thanks for the time and effort spent on it! :raised_hands:

**Comment by emtiu** [comment]
Thanks a lot for sharing! We're still getting settled in with a new maintenance team, but we'll take a look at what's going on, and how your personal hotfix may improve things for everyone :)

**Comment by buhtzz** [comment]
@eddgrant Awesome. Thank you so much for reporting and analysing this.

I'm not well experienced with virtual environments (e.g. `pyenv`) because I don't trust them. :D But I have an idea how they work. I will try to reproduce this in a VM.

If I understood you correct you run backintime in a virtuell environment? How can this happen? Or somehow the `sh` run in the environment?

btw: Maybe a `/usr/bin/env python3 -Es ${APP_PATH}/app.py "$@"` will help here.

@emtiu Didn't we had other issues with importing `PyQt5`?

**Comment by emtiu** [comment]
> @emtiu Didn't we had other issues with importing `PyQt5`?

From a quick search, that would be:
- no longer relevant https://github.com/bit-team/backintime/issues/1004
- different problem: https://github.com/bit-team/backintime/issues/1042

So, this one looks unique to me at first glance.

**Comment by buhtzz** [comment]
> I think what was happening in my case was that my [pyenv](https://github.com/pyenv/pyenv) configuration

How did you installed pyenv? Did you [build and installed](https://github.com/pyenv/pyenv#automatic-installer) it direct from github? It seems to me not installable via pip.

**Comment by eddgrant** [comment]
@buhtzz , here's the extra diagnostic info you asked for:

```shell
➜  ~ backintime-qt
/home/egrant/.pyenv/versions/3.9.9/bin/python3
Traceback (most recent call last):
  File "/usr/share/backintime/qt/app.py", line 35, in <module>
    import qttools
  File "/usr/share/backintime/qt/qttools.py", line 21, in <module>
    from PyQt5.QtGui import (QFont, QColor, QKeySequence)
ModuleNotFoundError: No module named 'PyQt5'
```

Looks like its using my pyenv default Python runtime (which it calls "global"), which is currently set to `3.9.9`:

```➜  ~ pyenv global
3.9.9
```

**Comment by eddgrant** [comment]
I think my pyenv installation probably pre-dates the automatic installer. IIRC I installed pyenv simply by cloning the git repo e.g.
`git clone https://github.com/pyenv/pyenv.git ~/.pyenv`

> How did you installed pyenv? Did you [build and installed](https://github.com/pyenv/pyenv#automatic-installer) it direct from github? It seems to me not installable via pip.

**Comment by eddgrant** [comment]
@buhtzz I think your understanding of how pyenv works sounds correct (at least to me!). As I understand it pyenv is initialised when a shell is started, which in turn executes its initialisation file (`.zsh`, `.bashrc` etc) which contains statements to initialise pyenv. Once pyenv is initialised it figures out which python runtime is desired (shell, local or global) and injects the relevant shims in to the environment.

**Comment by buhtzz** [comment]
Referring to your headline I would say it isn't the responsibility of an application to point to the correct (specific) interpreter. BIT says it needs Python version 3 and then take what the operating system (or shell) does offer.

In your case you kind of "hack" a bit and manage multiple python interpreters of different versions beside the operating systems itself. PyEnv "steal" the responsibility from the OS.

IMHO the reasons for that Issue is that you misused PyEnv. Use PyEnv to get the correct interpreter and its site-packages. Then it should work. Maybe PyEnv is able to point to the original OS own interpreter, too.

### Background

PyEnv isn't a virtual environment. It is just kind of a Python interpreter switcher enabling you to install multiple Python interpreters of different versions all with there own site-packages. You can switch between them. The selection is globally in the current shell valid. It is not restricted to a specific folder.

### My opinion

Because of that I vote for Close.

### Consequences

Despite that your report is valuable. It points me to an improvement for our diagnostics module.

**Comment by aryoda** [comment]
@eddgrant Excellent problem analysis - thanks a lot for sharing your knowledge here!

`pyenv` explicitly declares to be able to
> ... change the global Python version on a per-user basis.

I fully understand that manipulating the search path (like `virtenv` does here)

* has a major impact on **every** tools that calls `python3` without a hard-coded path to the OS-packaged version
  with  sometimes fatal results (like non-installed packages) - this is how Trojans try to get called ;-)
* has its uses cases (eg. when you want to intentionally inject a newer python version for testing purposes)

and that users doing this are fully responsible for the impact.

But: IMHO we should prevent starting the wrong (non-OS-packages) python version for **packaged** BiT versions to reduce support efforts.

I don't know how, but the OS package maintainers could possibly be able to solve this (it is a packaging issue IMHO).

@buhtzz How about discussing this eg. with the DEBIAN package maintainer of BiT as a starting point, AFAIR you have an account there. What do you think?

**Comment by aryoda** [comment]
I think @eddgrant opened this issue more with the intention to share his insights and less to urge a fix.

Nevertheless if I `grep -i python3 /usr/bin/*` I can find a lot of shebangs with hard coded python3 paths so I think
it could be solved by either using an absolute path injected by the package maintainer or a generic "find distro-packaged python" function in the starter script at our (BiT) side.

```
...
/usr/bin/asan_symbolize-9:#!/usr/bin/env python3
/usr/bin/backintime:python3 -Es $APP_PATH/backintime.py "$@"
/usr/bin/backintime-askpass:python3 -Es $APP_PATH/askpass.py "$@"
/usr/bin/backintime-qt:python3 -Es ${APP_PATH}/app.py "$@"
/usr/bin/chardet3:#! /usr/bin/python3
/usr/bin/chardetect3:#! /usr/bin/python3
/usr/bin/check-language-support:#! /usr/bin/python3
/usr/bin/chrome-gnome-shell:#!/usr/bin/python3
...
```

I think a simple `if` in the shell script could heavily improve the situation, eg. (not tested!):

```
# ...
python_starter="/usr/bin/python3"
if [ ! -f $python_starter ]; then
    python_starter="python3"
    echo "Unexpected python3 installation path detected (virtual env activated?). BiT may not work as expected"
fi

$python_starter -Es ~/dev/backintime/qt/app.py "$@"
```

**Edit:** If this shouldn't use the correct python version the package maintainers of BiT will come back to us anyhow ;-)

**Edit 2:** BiT common (CLI) is also affected by this issue...

**Edit 3:** Should we also add the path to the active python file to our diagnostics output (`which python3`)?

**Comment by buhtzz** [comment]
> I assume that @eddgrant has some problems to understand how `pyenv` should be used.

@eddgrant I would like to apologize for this. After I had read that again, it struck me that this could be understood very impolitely. It was not meant that way and can possibly be explained with my rather rough English. Although the latter may also not always be the excuse.

**Comment by aryoda** [comment]
OK, so let's settle this with "watch status" (don't change anything for now)
to observe if more issues arise due to non-standard python installations.

The new `--diagnostics` feature with your PR #1318 to show the active python should allow us to diagnose these types of problems in issues now...

**Comment by emtiu** [comment]
Thanks for a thorough discussion, everyone. And thanks @buhtzz for explaining to @eddgrant that you meant no insult :)

I agree with @buhtzz that it's a sane default for a caller script to simply call `python3`. Where does it lead? That's defined by the distro/system first, the system admin second, and the user third. If any of them change the destination of `python3` by modifying `$PATH` or inserting symlinks, then it's their responsibility to check that python software still works as expected.

We _might_ put `/usr/bin/python3` in there, with `python3` as a fallback, and it probably wouldn't break things (though I'm not willing to bet my bike on it ;)). But I don't think it's necessary, since really, a system/user should always have `python3` point to somewhere that makes sense. I also happen to know that FreeBSD will always have `/usr/local/bin/python3` instead of `/usr/bin/python3`, so that would be a "miss".

I also agree with @aryoda that we should keep this Issue open for reference. Python is a lively language, and conditions might change. Again, thanks @eddgrant for sharing!

**Comment by aryoda** [comment]
@emtiu Do you have an idea for a new label with a self-explanatory name ("watchlist" or better) for this type of issue status?

**Comment by emtiu** [comment]
IMHO, https://github.com/bit-team/backintime/labels/Discussion fits the bill well enough. Do you think we might add "Reference" for this case?

**Comment by aryoda** [comment]
@emtiu I would interpret "Discussion" as an active discourse or waiting for an answer to a question.
What I search is something like "**resubmission**" since the issue sleeps until it is waked up again due to a similar issue or finally closed after the resubmission period.

Let's keep "Discussion", it is a good filter to be **revisited** again from time to time...

**Comment by aryoda** [comment]
I just stumbled across the solution how to inject the OS-specific python path into all launcher scripts and configurations:

Via `./configure` in `common` and `qt`.

This is what package maintainers call when preparing a package (together with `make` and `make install` to be ready to package the installed files).

If we extend the existing logic to inject python2 vs. python3 by also injecting the absolute path (eg. found with `which python3`)
it should work universally (there may be a better way than `which` I may look into the clever logic of the another configuration tool (`autoconf` for C is very clever AFAIR):

https://github.com/bit-team/backintime/blob/b4dd2d1737c5031d264d667c18f6987a1916762b/common/configure#L133-L141

**Comment by emtiu** [comment]
That sounds like it should work. It would be a lot of effort to test, and we can't cover all bases, but it seems reasonable.

**Comment by aryoda** [comment]
Yes, too much efforts for now, let' stick to the `to be revisited again from time to time...` plan to learn how much other users are impacted at all

BTW: You will find me quite often posting "how it could be fixed" or at least "how does the current implementation work" comments even for non-HIGH issues because I am trying to understand the BiT full code base by digging into issues. This does not imply "must be fixed [at all or right now]" unless the fix is trivial and low-risk...

**Comment by aryoda** [comment]
FYI:

OpenSuse 15.3 uses this command in the installation script of the BiT package (v1.2.1):

https://build.opensuse.org/package/view_file/openSUSE:Leap:15.3/backintime/backintime.spec?expand=1
```
# Fix shebangs
sed -i 's|/usr/bin/env python|#!/usr/bin/python|g' common/askpass.py
```

**Comment by buhtz** [comment]
Let's review this issue after #1575 (migration to Python build system) is done.

**Comment by aryoda** [comment]
I think we should close this issue after a cooling-off period of a few weeks since we have merged PR #1602 to solve this:

It is possible to configure an absolute path to python3 with `./configure --python=<path>`.

By default we still use no absolute path for the BiT starter scripts...

**Comment by buhtz** [comment]
Closing this ticket based on the comment above. Feel free to reopen
if the problem is still relevant. Thank you for your efforts. If you have
any further questions, ideas or encounter any other issues, please
don't hesitate to let us know.

Best regards,
